### ISSUE TITLE
`Model.predict()` generates unexpected out-of-sample predictions for a mixed effects model

### ISSUE THREAD

**Issue by LeonieMei** [bambinos/bambi#477]
Hello,

first of all, thanks for the great work on this project, I've been using bambi a lot and it has been super helpful!

I'm currently facing a (potential) issue when trying to make out-of-sample predictions for a logistic regression model built with the following formula:

`y ~ x1 + x2 + x3 + (0 + x2|x1) + (0 + x3|x1) `

where `x1` and `x2` are categorical variables with two dimensions respectively and `x3` is a continuous variable.

The out-of-sample data I'm trying to make predictions for looks like this (exemplary):

| x1 | x2 | x3  |
|----|----|-----|
| 0  | 0  | 0   |
| 0  | 0  | 0.5 |
| 0  | 0  | 1   |
| 0  | 0  | 1.5 |
| 1  | 0  | 0   |
| 1  | 0  | 0.5 |
| 1  | 0  | 1   |
| 1  | 0  | 1.5 |

There was no error when running `model.predict(iData, data=out_of_sample_data, kind='mean')`, however the spaghetti plot I generated from the posterior predictions looked off for when `x1==1`, the variance was much bigger than I expected. (I noticed this because I had manually made a plot displaying the 0.5 decision boundary for `x3`, i.e. the mean value and hdi intervals of `x3` where there is a 50% probability of a positive outcome and that didn't match what I saw in the spaghetti plot.)

I then had a look at the code and noticed that the `Z` matrix generated in the `predict` method in `models.py` looked different from what I expected.
Here's the code bit I'm referring to (last line):
```python
        if self._design.group:
            if in_sample:
                Z = self._design.group.design_matrix
            else:
                Z = self._design.group._evaluate_new_data(data).design_matrix
```

What I got for `Z` was the following:

|||||||
|---|---|---|---|-----|-----|
| 1 | 0 | 0 | 0 | 0   | 0   |
| 1 | 0 | 0 | 0 | 0.5 | 0   |
| 1 | 0 | 0 | 0 | 1   | 0   |
| 1 | 0 | 0 | 0 | 1.5 | 0   |
| 0 | 0 | 1 | 0 | 0   | 0   |
| 0 | 0 | 1 | 0 | 0   | 0.5 |
| 0 | 0 | 1 | 0 | 0   | 1   |
| 0 | 0 | 1 | 0 | 0   | 1.5 |

...but what I was expecting (after trying to make sense of it) was this:

|||||||
|---|---|---|---|-----|-----|
| 1 | 0 | 0 | 0 | 0   | 0   |
| 1 | 0 | 0 | 0 | 0.5 | 0   |
| 1 | 0 | 0 | 0 | 1   | 0   |
| 1 | 0 | 0 | 0 | 1.5 | 0   |
| 0 | 1 | 0 | 0 | 0   | 0   |
| 0 | 1 | 0 | 0 | 0   | 0.5 |
| 0 | 1 | 0 | 0 | 0   | 1   |
| 0 | 1 | 0 | 0 | 0   | 1.5 |

so basically the second and third column swapped. I then added the following line:

```python
Z[:, [1, 2]] = Z[:, [2, 1]]
```

to achieve that and the spaghetti plot I then generated matched my expectation.

It would be great if someone had a look at this and fixed it properly (if it really is an issue and not me making a mistake), I hope it was clear enough and if not, let me know!

**Comment by tomicapretto** [comment]
Hi @LeonieMei 

Thanks for reporting this issue and the investigation you carried out to understand what was going on. It looks like a problem indeed.

Can I ask you for a reproducible example so we are both on the same page? Random generated data that looks like your data is enough. Or even an anonymized sample of your dataset would work if you're able to do that.

Thanks!

**Comment by LeonieMei** [comment]
Thanks a lot for the quick reply! I simulated some data (`data.tsv`) and am also sending the out-of-sample data I used for making predictions (`dataNew.tsv`). The figures show the spaghetti plot for when I'm using the original bambi code (`prediction1.png`) vs when using the code with the adaptation I've mentioned above (`prediction2.png`).

[data_and_plots.zip](https://github.com/bambinos/bambi/files/8343857/data_and_plots.zip)

Here's how I built and fit the model:

```python
formula = 'y ~ x1 + x2 + x3 + (0 + x2|x1) + (0 + x3|x1)'
priors = {'x2|x1': bmb.Prior('Normal', mu=0, sigma=bmb.Prior('HalfNormal', sigma=1)),
               'x3|x1': bmb.Prior('Normal', mu=0, sigma=bmb.Prior('HalfNormal', sigma=1)),
               'x3': bmb.Prior('Lognormal', mu=1, sigma=1),}
model = bmb.Model(data=df, formula=formula, family='bernoulli', categorical=['x1', 'x2'], priors=priors)
model.build()
iData = model.fit(target_accept=0.95)
```

Thanks again!

**Comment by tomicapretto** [comment]
I've been trying to understand this problem. All the design matrix stuff happens in formulae, not in Bambi. So let's use formulae to understand what's going on.

Setup...
```python
import pandas as pd
from formulae import design_matrices

df = pd.read_csv("data.tsv", sep='\t')
data_new = pd.read_csv("dataNew.tsv", sep="\t")

formula = "y ~ x1 + x2 + x3 + (0 + x2|x1) + (0 + x3|x1)"
categorical = ["x1", "x2"]
```

First, we need to note what Bambi does with the `categorical` argument.

```python
data = df.copy()
cats = categorical
data[cats] = data[cats].apply(lambda x: x.astype("category"))
```

Now let's create the design matrix object.

```python
dm = design_matrices(formula, data)
dm.group
```

```
GroupEffectsMatrix(  
  shape: (300, 6),
  terms: {  
    'x2|x1': {  
      type: categoric,
      groups: ['0', '1'],
      levels: [0, 1],
      reference: 0,
      encoding: full,
      full_names: ['x2[0]|0', 'x2[1]|0', 'x2[0]|1', 'x2[1]|1'],
      cols: slice(0, 4, None)
    },
    'x3|x1': {  
      type: numeric,
      groups: ['0', '1'],
      full_names: ['x3|x1[0]', 'x3|x1[1]'],
      cols: slice(4, 6, None)
    }
  }
)
```
The output above is the one you would obtain with `formulae==0.2.0`, which is the one that Bambi installs. The last release is `formulae==0.3.0`, which changes a lot of things, but it's not used by Bambi yet. Fortunately you won't notice any change on Bambi side when I implement the switch.

We focus on the term `'x2|x1`, which is the one causing troubles in your analysis. Let's have a look at the following

```
full_names: ['x2[0]|0', 'x2[1]|0', 'x2[0]|1', 'x2[1]|1']
```

This is telling that the first column corresponds to `x2==0` and `x1==0`, the second to `x2==1` and `x1==0`, the third to `x2==0` and `x1==1` and the fourth to `x2==1` and `x1==1`.

Let's now predict (i.e generate a new design matrix) from the new dataset. To save some space I'm going to use unique combinations of `x2` and `x1` in the data.

```python
data_new_unique = data_new.drop_duplicates(["x1", "x2"])
data_new_unique
```

|    |   x1 |   x2 |   x3 |
|---:|-----:|-----:|-----:|
|  0 |    0 |    0 |   -2 |
| 80 |    1 |    0 |   -2 |


And now let's see the new matrix (which is `Z` in Bambi's code for predictions)

```python
dm_new = dm.group._evaluate_new_data(data_new_unique)
dm_new["x2|x1"]
```
```
array([[1., 0., 0., 0.],
       [0., 0., 1., 0.]])
```

In the first row, we have `x1==0` and `x2==0`. That means the first column is turned on. 
In the second row, we have `x1==1` and `x2==0`. That means the third columns is turned on.

I think this is exactly what is happening in your predictions, and I don't think there's nothing wrong here (at least in the generation of the design matrices).

If you double check and you're convinced the results are still wrong, we may need to check the underlying PyMC3 variables, to see if the order there matches the order of the columns in Z. As far as I am aware, they match, but maybe we're facing a more complicated bug.

**Comment by LeonieMei** [comment]
Thanks a lot for your extensive reply and digging into it! What you write makes total sense and I wouldn't say I'm convinced that the results are wrong but there is one thing I find odd. I guess it's rather that I think the `Z` matrix and `beta_z` don't fit together, see this line in `models.py`:

```python
            contribution = np.dot(Z, beta_z.T).T
```
When I'm printing the following after fitting the model,

```python
iData.posterior.sel( x2_coord_group_expr=0, x1_coord_group_factor='1',)['x2|x1']
```
i.e. `x2=0` and `x1=1`, I get the values that are in the second column of `beta_z.T`. However, according to your explanation, they should be in the third column so they are multiplied with the corresponding new data values in the `Z` matrix. I hope that makes sense.

**Comment by tomicapretto** [comment]
I confirm I can reproduce what you are reporting. 

As far as I understand now, the problem is in the reshape. Let's see the following investigation.

Let's see the first posterior draw of the first chain.

```python
posterior = iData.posterior["x2|x1"]
posterior.sel(chain=0, draw=0)
```

![image](https://user-images.githubusercontent.com/25507629/160386702-041b241d-af58-4d5e-a58f-77ffe3c7489a.png)

If we look at the array values as a matrix, we see that 'expr' varies along the rows, and 'factor' varies along the columns.
We can confirm with the following

```python
posterior.sel(chain=0, draw=0, x2_coord_group_expr=0, x1_coord_group_factor='0').values
```
```python
array(-1.02444546)
```

```python
posterior.sel(chain=0, draw=0, x2_coord_group_expr=1, x1_coord_group_factor='0').values
```

```python
array(-0.7650879)
```

However, when we do (which is what is happening internally [here](https://github.com/bambinos/bambi/blob/284571538a1a8946404da8ee4e1726b52541fee7/bambi/models.py#L861))

```python
posterior.sel(chain=0, draw=0).values.reshape(4)
```

We see the following result

```python
array([-1.02444546, -1.20200069, -0.7650879 , -1.03656599])
```

So `array[0]` corresponds to `expr==0` and `factor==0` but `array[1]` corresponds to `expr==1` and `factor==1`. So what varied first was the second axis, and not the first one. 

I think that is right according to Numpy's documentation on the matter https://numpy.org/doc/stable/reference/generated/numpy.reshape.html.

Anyway, **this bug affects models with categorical group-specific effects**.

Now I need to find a solution :smile:

**Comment by LeonieMei** [comment]
Ok yes, that makes sense. Just to be sure, I guess you meant `array[1]` corresponds to `expr==0` and `factor==1`? 
As for a solution, I just checked and if you write `posterior.sel(chain=0, draw=0).values.reshape(4, order='F')` you get the right order but I don't know if this applies to other use cases as well.

**Comment by tomicapretto** [comment]
Yes, you are right, it was a typo in my previous comment. 

Using that "F" would work for that particular case where we subset chain and draw, but it's not going to work as a general rule. 

I'll try to propose a solution in the coming days.

**Comment by LeonieMei** [comment]
Okay great, thanks a lot!

**Comment by tomicapretto** [comment]
@LeonieMei I opened a #478 where I try to fix this. Looks like reshaping in two steps is the solution.

Could you install from that PR and check if the results make sense on your side now?

**Comment by LeonieMei** [comment]
Great, yes the plots look fine! Thanks again for the quick help and fix! 🥳

**Comment by terrycojones** [comment]
This is great :-)  Thanks @tomicapretto!

**Comment by tomicapretto** [comment]
Cool! I need to test if this works with more cases. 
If you have other examples where you have categorical group-specific effects you should use this updated code because the previous one was just wrong. Also, if you can, let me know if those tests run OK so I have more data about this patch.

**Comment by LeonieMei** [comment]
Yes, I also made a plot for a model with basically the same specification but instead of two categories in the group variable I had four and it also worked in that case 👍

**Comment by tomicapretto** [comment]
#478 is now merged. Thanks for reporting the issue!

**Comment by LeonieMei** [comment]
Another thing I noticed is that the output when fitting the model is a bit off:
<img width="302" alt="Bildschirmfoto 2022-04-25 um 16 38 18" src="https://user-images.githubusercontent.com/71067488/165112157-cc6ee99b-2eb6-4522-8448-906e9f75350a.png">

**Comment by tomicapretto** [comment]
@LeonieMei thank you so much for this input. 

I was using `.stack(coefs=tuple(coord_expr + coord_factor))` but actually I needed `.stack(coefs=tuple(coord_factor + coord_expr))`. The thing is that first we need to vary along the expr values and then along factor values. 

I've repeated the analysis on my side and results look good now. It would be good if you could add your input too, but I'm happy with the help so far.

About the second issue, I've fixed it too.
